### PR TITLE
New Python interface, new TREs, other enhancements

### PR DESCRIPTION
**Pull request by tclarke**
This code is not ready to merge, there needs to be some documentation, cleanup, etc. I wanted to make the merge request to start a discussion on what needs to be changed for merge consideration. A summary of changes:

- The biggest change is a new cython based Python 3 compatible wrapper. It's originally based on the general design and layout of the existing SWIG wrapper but also has some more pythonic interfaces. (the old stuff is marked deprecated). I've been using it pretty extensively on OSX and Centos. I've built and done some basic tests on Windows 7/10 and it seems to be working there. `modules/python/cython' contains the code. It's expecting nitro to be in `../../../install`. If it isn't, then change the setup.py appropriately.
- There are a few newer TREs included, mostly from the upcoming NGA SNIP profile.
- There's some changes to the ImageWriter to try and allow for direct writes when there's more than one band. Specifically, I'm trying to optimize writes of large (multiple GB) BIP images. I'm not sure if this is working completely right now and I'll probably pull it out of an initial merge.

### CODE REVIEW

File: modules/python/cython/field.pyx [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            if not PyCapsule_IsValid(capsule, "Field"):
+                raise TypeError("Invalid C pointer type")
+            self._c_field = <field.nitf_Field*>PyCapsule_GetPointer(capsule, "Field")
+        elif length is not None or value is not None:
+            if length is None:
+                length = len(value)
+            self._c_field = nitf_Field_construct(length, field_type, &error)
+            if self._c_field is NULL:
+                raise NitfError(error)
+            if value is not None:
```
Comment (JonathanMeans) [inline modules/python/cython/field.pyx]: I'm concerned that a policy of making the caller decide whether they have to manually set the type could lead to annoying bugs if someone forgets. But then it looks like the only cases where something could go wrong are with file length and "Date and Time" fields. Could there be some warning in the .pxd file (I take it that's like a .h for cython?) about these particular cases being manual?

File: modules/python/cython/field.pyx [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        return FieldType(self._c_field.type)
+
+    def __len__(self):
+        return self._c_field.length
+
+    def __str__(self):
+        if self.type in [NITF_BCS_N, NITF_BCS_A]:
+            return self.get_string()
+        # this generates a nice list of hex byte values truncating at 200 bytes like numpy array pretty printing
+        with printoptions(threshold=200, edgeitems=100, formatter={'int_kind':lambda v: '%02x' % v}):
```
Comment (JonathanMeans) [inline modules/python/cython/field.pyx]: This is really cool.

File: modules/python/cython/test_des.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import nitro
```
Comment (JonathanMeans) [inline modules/python/cython/test_des.py]: Could you put these in `modules/python/cython/unittests`? (And hopefully they end up getting installed into `install/unittests/cython`) I have a script that picks up everything in `install/unittests` and automatically runs them.

Comment (JonathanMeans) [reply modules/python/cython/test_des.py]: Wait, it looks like the caller has to manually check the file got created and clean up after, so in this case it wouldn't belong in `unittests`

**Comment by tclarke** [comment]
Haven't forgotten about this...got a couple more changes in the works before I update the request.

**Pull request closed without merging**